### PR TITLE
Add top level docs

### PR DESCRIPTION
**Pull request by blink1073**
cc @jasongrout

### CODE REVIEW

**Comment by blink1073** [comment]
cc @willingc

File: README.md [hunk 1]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+By using a semver range, we enable client-side deduplication of modules, where 
+the registered module that maximally satisfies a semver range is the one 
+returned by the `require` function call.  This also enables us to perform 
+server-side deduplication of modules prior to serving the bundles, the 
+client-side lookup will still load the correct modules.  Reasons to deduplicate
+code include being able to use `isinstance()` on an object to determine if
+it is the same class (a technique used by phosphor's drag-drop mechanism).  
+Another reason is for sharing of module-private state between different
+consumers, such as a list of client-side running kernels in 
+`jupyter-js-services.
```
Comment (jasongrout) [inline README.md]: missing backtick

File: README.md [hunk 2]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
 
 Tools for building JupyterLab extensions.
 
+An extension to JupyterLab is a module that provides one or more
+plugins to the JupyterLab application.  This library provides
```
Comment (willingc) [inline README.md]: A JupyterLab extension provides additional, optional functionality to JupyterLab's built-in capabilities. The extension is a module that provides one or more plugins to the JupyterLab application. To streamline third-party development of extensions, this library...

File: README.md [hunk 3]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+a build script for generating third party extension JavaScript
+bundles.  
+
+Simple extensions can be created by using the `buildExtension` function
+with the default options.  More advanced extensions may require additional
+configuration.
```
Comment (willingc) [inline README.md]: configuration, such as ????, ???, etc.

File: README.md [hunk 4]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+
+A simple extension entry point could look like:
```
Comment (willingc) [inline README.md:16]: entrypoint (?)

File: README.md [hunk 5]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+```javascript
+module.exports = [{
+    id: 'my-cool-extension',
+    activate: function(app) {
+       console.log(app.commands);
+    }
+}];
+```
+
+The above could be built using the following script:
```
Comment (willingc) [inline README.md]: Build the above example...

File: README.md [hunk 6]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+var buildExtension = require('jupyterlab-extension-builder').buildExtension;
+
+buildExtension({
+    name: 'my-cool-extension',
+    entryPath: './index.js',
+    outputPath: './build'
+});
+```
+
+In this case the builder script will create the following files:
```
Comment (willingc) [inline README.md]: files in the `build` directory:

File: README.md [hunk 7]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+
+```
+build/my-cool-extension.js
+build/my-cool-extension.js.manifest
+```
+
+Other extensions may produce other files in the build directory
```
Comment (willingc) [inline README.md]: produce ~~other~~ additional files in the `build`

File: README.md [hunk 8]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+depending on the complexity of extension.  The two files above
```
Comment (willingc) [inline README.md]: above, `my-cool-extension.js` and `my-cool-extension.js.manifest`,

File: README.md [hunk 9]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+are used by JupyterLab server to determine the entry point file(s) and
```
Comment (willingc) [inline README.md]: ...by the JupyterLab...

Not sure whether it should be entry point or entrypoint. Either I suspect is fine as long as we are consistent.

File: README.md [hunk 10]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+build/my-cool-extension.js
+build/my-cool-extension.js.manifest
+```
+
+Other extensions may produce other files in the build directory
+depending on the complexity of extension.  The two files above
+are used by JupyterLab server to determine the entry point file(s) and 
+entry point module(s) for the extension.  The extension must also be 
+registered using `jupyter labextension` in order to be added to the 
+application.
```
Comment (willingc) [inline README.md]: The extension must also be registered, using the command `jupyter labextension`, in order to be added to the JupyterLab application.

**Review by willingc — COMMENTED** [review]
First half of the review. Looks really good. Some suggestions for clarity for those not familiar with the process.

File: README.md [hunk 11]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+
+The bundles are created using Webpack, where the modules produced
+by Webpack are modified to use the JupyterLab custom module registration
+and loading mechanism.
```
Comment (willingc) [inline README.md]: The extension bundles are created using Webpack, and the modules produced by Webpack are modified to use JupyterLab's custom module registration and loading mechanism.

File: README.md [hunk 12]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+are used by JupyterLab server to determine the entry point file(s) and 
+entry point module(s) for the extension.  The extension must also be 
+registered using `jupyter labextension` in order to be added to the 
+application.
+
+The bundles are created using Webpack, where the modules produced
+by Webpack are modified to use the JupyterLab custom module registration
+and loading mechanism.  
+
+The module loading mechanism uses a `define` function that registers
```
Comment (willingc) [inline README.md]: JupyterLab's custom **module registration and loading mechanism** uses...

File: README.md [hunk 13]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+modules by name, where the name contains the package name, version 
+number, and the full path to the module.  For example, 'phosphor@0.6.1/lib/ui/widget.js'.  Within a `define` function, modules are required by package
```
Comment (willingc) [inline README.md]: ... a required module is referenced by ...

File: README.md [hunk 14]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+name, semver range, and the full path to the module.  For example,
+`require('phosphor@^0.6.0/lib/ui/tabpanel.js')`.  
+
+By using a semver range, we enable client-side deduplication of modules, where
```
Comment (willingc) [inline README.md]: By using a server range, JupyterLab can perform client-side deduplication of modules, where the registered module ...

File: README.md (hunk 10 above)
Comment (blink1073) [reply README.md]: Great feedback, thanks @willingc, all comments addressed.

File: README.md [hunk 15]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+The module loading mechanism uses a `define` function that registers
+modules by name, where the name contains the package name, version 
+number, and the full path to the module.  For example, 'phosphor@0.6.1/lib/ui/widget.js'.  Within a `define` function, modules are required by package
+name, semver range, and the full path to the module.  For example,
+`require('phosphor@^0.6.0/lib/ui/tabpanel.js')`.  
+
+By using a semver range, we enable client-side deduplication of modules, where 
+the registered module that maximally satisfies a semver range is the one 
+returned by the `require` function call.  This also enables us to perform 
+server-side deduplication of modules prior to serving the bundles, the
```
Comment (willingc) [inline README.md]: bundles, and the ...

File: README.md [hunk 16]
```diff
@@ -3,6 +3,85 @@ JupyterLab Extension Builder
...
+By using a semver range, we enable client-side deduplication of modules, where 
+the registered module that maximally satisfies a semver range is the one 
+returned by the `require` function call.  This also enables us to perform 
+server-side deduplication of modules prior to serving the bundles, the 
+client-side lookup will still load the correct modules.  Reasons to deduplicate
+code include being able to use `isinstance()` on an object to determine if
+it is the same class (a technique used by phosphor's drag-drop mechanism).  
+Another reason is for sharing of module-private state between different
+consumers, such as a list of client-side running kernels in 
+`jupyter-js-services`.
```
Comment (willingc) [inline README.md]: Reasons to deduplicate code include:
-  being able to use `isinstance()` on an object to determine if it is the same class (a technique used by phosphor's drag-drop mechanism)
-   sharing of module-private state between different consumers, such as a list of client-side running kernels in `jupyter-js-services`.

File: README.md [hunk 17]
```diff
@@ -3,6 +3,87 @@ JupyterLab Extension Builder
...
+
+All client-side `require()` calls are synchronous, which means that the 
+bundles containing the `define()` modules must be loaded prior to using
+any of their functions.  The loader provides an `ensureBundle()` function
```
Comment (willingc) [inline README.md]: ~~their~~ the bundles'

File: README.md [hunk 18]
```diff
@@ -3,6 +3,87 @@ JupyterLab Extension Builder
...
+to load a particular bundle or bundles prior to calling `require()` on
+a module.
+
+There may be some cases where the `buildExtension` function is not sufficient
+to build an extension and a completely custom Webpack configuration is
```
Comment (willingc) [inline README.md]: A _completely custom Webpack configuration_ may be needed if there is a case where the `buildExtension` function is not sufficient to build the extension. If a custom Webpack configuration is needed, the `JupyterLabPlugin` must be used as part of the Webpack config to ensure proper handling of module definition and requires.

**Review by willingc — COMMENTED** [review]
Second half of review finished. Nice!

**Comment by willingc** [comment]
@blink1073 :heart: documentation. Thanks to you and @jasongrout. Use whichever suggestions work for you, and then :shipit:

**Comment by blink1073** [comment]
Thanks again, updated!

**Comment by willingc** [comment]
Pressing the green button. We can iterate as needed @jasongrout.

**Pull request merged**